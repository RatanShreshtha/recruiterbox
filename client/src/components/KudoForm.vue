<template>
  <b-form @submit="onSubmit">
    <b-form-group id="input-group-1" label="User:" label-for="input-1">
      <b-form-select
        id="input-1"
        v-model="user"
        :options="users"
        required
      ></b-form-select>
    </b-form-group>

    <b-form-group id="input-group-2" label="Title:" label-for="input-2">
      <b-form-input
        id="input-2"
        v-model="title"
        type="text"
        required
        placeholder="Enter title"
      ></b-form-input>
    </b-form-group>

    <b-form-group id="input-group-3" label="description:" label-for="input-3">
      <b-form-textarea
        id="input-3"
        v-model="description"
        placeholder="Enter description..."
        rows="7"
        max-rows="10"
      ></b-form-textarea>
    </b-form-group>

    <b-button class="mr-2" type="submit" variant="primary">Give Kudo</b-button>
  </b-form>
</template>
<script>
import router from "@/router";
import swal from "sweetalert";
import { mapState } from "vuex";

import { apiService } from "@/service";

export default {
  data() {
    return {
      user: null,
      title: "",
      description: "",
      message: null,
      error: null,
      users: [{ text: "Select User to give kudos.", value: null }]
    };
  },
  computed: {
    ...mapState(["id", "first_name", "last_name", "username", "token", "kudos"])
  },
  async mounted() {
    this.getUsers();
  },
  methods: {
    async getUsers() {
      try {
        let nextPage = "/api/users/";

        while (nextPage !== null) {
          const resp = await apiService.get(nextPage, {
            headers: {
              Authorization: `Token ${this.token}`
            }
          });
          nextPage = resp.data.next;
          for (let userData of resp.data.results) {
            const userObj = {
              text: `${userData.first_name} ${userData.last_name}`,
              value: `${userData.id}`
            };
            this.users.push(userObj);
          }
        }
      } catch (error) {
        console.log("error", error);
      }
    },
    async onSubmit(evt) {
      evt.preventDefault();
      try {
        const kudoData = {
          title: this.title,
          description: this.description,
          by_user: this.id,
          to_user: this.user
        };
        console.log(kudoData);

        await apiService.post("/api/kudos/", kudoData, {
          headers: {
            Authorization: `Token ${this.token}`
          }
        });

        swal(
          "Kudo is away ...",
          `You now have ${this.kudos - 1} left.`,
          "success"
        ).then(() => {
          this.$store.dispatch("setKudos", this.kudos - 1);
          router.replace("/kudos-given");
        });
      } catch (error) {
        console.log("error", error);
        this.error = error.message;
      }
    }
  }
};
</script>

<style></style>
