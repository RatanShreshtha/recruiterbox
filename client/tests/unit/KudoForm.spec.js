import { shallowMount, createLocalVue } from "@vue/test-utils";

import { createRenderer } from "vue-server-renderer";

import Vuex from "vuex";
import VueRouter from "vue-router";
import { BootstrapVue, IconsPlugin } from "bootstrap-vue";
import createPersistedState from "vuex-persistedstate";

import KudoForm from "@/components/KudoForm.vue";

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueRouter);

localVue.use(BootstrapVue);
localVue.use(IconsPlugin);

describe("KudoForm.vue", () => {
  let state;
  let store;

  beforeEach(() => {
    state = {
      id: null,
      kudos: 0,
      username: null,
      first_name: null,
      last_name: null,
      email: null,
      token: null
    };

    store = new Vuex.Store({
      plugins: [createPersistedState()],
      state
    });
  });
  it("matches snapshot", () => {
    const renderer = createRenderer();

    const wrapper = shallowMount(KudoForm, {
      localVue,
      store
    });

    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err);
      expect(str).toMatchSnapshot();
    });
  });
});
