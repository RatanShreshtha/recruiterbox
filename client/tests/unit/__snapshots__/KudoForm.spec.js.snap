// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KudoForm.vue matches snapshot 1`] = `
<b-form-stub>
  <b-form-group-stub id="input-group-1" label="User:" labelFor="input-1" feedbackAriaLive="assertive">
    <b-form-select-stub id="input-1" required="required" options="[object Object]" valueField="value" textField="text" htmlField="html" disabledField="disabled" labelField="label" optionsField="options" selectSize="0"></b-form-select-stub>
  </b-form-group-stub>
  <b-form-group-stub id="input-group-2" label="Title:" labelFor="input-2" feedbackAriaLive="assertive">
    <b-form-input-stub id="input-2" required="required" value="" placeholder="Enter title" debounce="0" type="text"></b-form-input-stub>
  </b-form-group-stub>
  <b-form-group-stub id="input-group-3" label="description:" labelFor="input-3" feedbackAriaLive="assertive">
    <b-form-textarea-stub id="input-3" value="" placeholder="Enter description..." debounce="0" rows="7" maxRows="10" wrap="soft"></b-form-textarea-stub>
  </b-form-group-stub>
  <b-button-stub target="_self" event="click" routerTag="a" variant="primary" type="submit" tag="button" class="mr-2">Give Kudo</b-button-stub>
</b-form-stub>
`;
